enterprise/dbcrypt/dbcrypt.go

```diff
@@ -0,0 +1,296 @@
+	if err != nil {
+		return database.GitAuthLink{}, err
+	}
+	return link, db.decryptFields(&link.OAuthAccessToken, &link.OAuthRefreshToken)
+}
+
+func (db *dbCrypt) GetGitAuthLinksByUserID(ctx context.Context, userID uuid.UUID) ([]database.GitAuthLink, error) {
+	links, err := db.Store.GetGitAuthLinksByUserID(ctx, userID)
+	if err != nil {
+		return nil, err
+	}
+	for _, link := range links {
+		if err := db.decryptFields(&link.OAuthAccessToken, &link.OAuthRefreshToken); err != nil {
+			return nil, err
+		}
+	}
+	return links, nil
+}
+
+func (db *dbCrypt) UpdateGitAuthLink(ctx context.Context, params database.UpdateGitAuthLinkParams) (database.GitAuthLink, error) {
+	err := db.encryptFields(&params.OAuthAccessToken, &params.OAuthRefreshToken)
+	if err != nil {
+		return database.GitAuthLink{}, err
+	}
+	updated, err := db.Store.UpdateGitAuthLink(ctx, params)
+	if err != nil {
+		return database.GitAuthLink{}, err
+	}
+	return updated, db.decryptFields(&updated.OAuthAccessToken, &updated.OAuthRefreshToken)
+}
+
+func (db *dbCrypt) SetDBCryptSentinelValue(ctx context.Context, value string) error {
+	err := db.encryptFields(&value)
+	if err != nil {
+		return err
+	}
+	return db.Store.SetDBCryptSentinelValue(ctx, value)
+}
+
+func (db *dbCrypt) encryptFields(fields ...*string) error {
+	// If no cipher is loaded, then we can't encrypt anything!
+	if db.ciphers == nil {
+		return ErrNotEnabled
+	}
+
+	for _, field := range fields {
+		if field == nil {
+			continue
+		}
+
+		encrypted, err := db.ciphers.Encrypt([]byte(*field))
+		if err != nil {
+			return err
+		}
+		// Base64 is used to support UTF-8 encoding in PostgreSQL.
+		*field = MagicPrefix + b64encode(encrypted)
+	}
+	return nil
+}
+
+// decryptFields decrypts the given fields in place.
+// If the value fails to decrypt, sql.ErrNoRows will be returned.
+func (db *dbCrypt) decryptFields(fields ...*string) error {
+	if db.ciphers == nil {
+		return ErrNotEnabled
+	}
+
+	for _, field := range fields {
+		if field == nil {
+			continue
+		}
+
+		if len(*field) < 8 || !strings.HasPrefix(*field, MagicPrefix) {
+			// We do not force decryption of unencrypted rows. This could be damaging
+			// to the deployment, and admins can always manually purge data.
+			continue
+		}
+
+		data, err := b64decode((*field)[8:])
+		if err != nil {
+			// If it's not base64 with the prefix, we should complain loudly.
+			return &DecryptFailedError{
+				Inner: xerrors.Errorf("malformed encrypted field %q: %w", *field, err),
+			}
+		}
+		decrypted, err := db.ciphers.Decrypt(data)
+		if err != nil {
+			// If the encryption key changed, return our special error that unwraps to sql.ErrNoRows.
+			return &DecryptFailedError{Inner: err}
+		}
+		*field = string(decrypted)
+	}
+	return nil
+}
+
+func ensureEncrypted(ctx context.Context, dbc *dbCrypt) error {
+	return dbc.InTx(func(s database.Store) error {
+		val, err := s.GetDBCryptSentinelValue(ctx)
+		if err != nil {
+			if !errors.Is(err, sql.ErrNoRows) {
```

Comment:
It's hard to figure out what the logic here is supposed to be.  At the lowest level there are several possibilities

1. There really is no row in the table
2. There really was a row, but it was not encrypted
3. There really was a row, it was encrypted, but we don't have a key to decrypt
4. There really was a row, it was encrypted, and we do have a key to decrypt

Making the decryption error unwrap to `ErrNoRows` means 1&3 maybe look the same?
Making the decryption transparent to unencrypted data means 2&4 look the same?

Are these cases really equivalent?

Reply:
3) is the main thing I want to guard against with this.

For both 1) and 2), my view is that we would want to update the row with an encrypted value. As we always write the same value (with ON CONFLICT UPDATE), I would view these two cases as equivalent.

I will admit that updating the value in case of 4) is technically not required, but I don't see any significant downside to it.